File: SteamDroid/src/com/flyingpie/steamdroid/util/SteamAlerts.java
package com.flyingpie.steamdroid.util;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.media.MediaPlayer;
import android.os.Vibrator;
import android.preference.PreferenceManager;
import android.util.Log;

import com.flyingpie.steamdroid.R;

/**
 * SteamAlerts handles notifications, sounds and vibrate
 * @author Marco vd Oever
 *
 */
public class SteamAlerts {

	private static MediaPlayer mediaPlayer;
	private static NotificationManager manager;
	private static Vibrator vibrator;
	
	private static Context context;
	
	private static ProgressDialog dialog;
	
	/**
	 * Initializes SteamAlerts, using the specified activity
	 * @param activity
	 */
	public static void initialize(Activity activity)
	{
		SteamAlerts.mediaPlayer = MediaPlayer.create(activity.getApplicationContext(), R.raw.message);
		SteamAlerts.manager = (NotificationManager)activity.getSystemService(Context.NOTIFICATION_SERVICE);
		SteamAlerts.vibrator = (Vibrator)activity.getSystemService(Context.VIBRATOR_SERVICE);
		
		SteamAlerts.context = activity.getApplicationContext();
	}
	
	/**
	 * Returns the context to use for the alerts
	 * @return
	 */
	public static Context getContext()
	{
		return context;
	}
	
	/**
	 * Shows a progress dialog with the specified parameters
	 * @param title
	 * @param message
	 * @param context
	 */
	public static void showProgressDialog(String title, String message, Context context)
	{
		if(dialog == null)
		{
			dialog = new ProgressDialog(context);
		}
		
		dialog.setTitle(title);
		dialog.setMessage(message);
		dialog.show();
	}
	
	/**
	 * Hides the progress dialog
	 */
	public static void hideProgressDialog()
	{
		if(dialog != null)
		{
			dialog.hide();
		}
	}
	
	/**
	 * Shows an alert with the specified parameters
	 * @param title
	 * @param message
	 * @param context
	 */
	public static void showAlert(String title, String message, Context context)
	{
		hideProgressDialog();
		
		AlertDialog.Builder alertUsername = new AlertDialog.Builder(context);
        alertUsername.setTitle(title);
        alertUsername.setMessage(message);
        alertUsername.setPositiveButton("Ok", new DialogInterface.OnClickListener() {
			@Override
			public void onClick(DialogInterface dialog, int which) {
				dialog.cancel();
			}
		});
        alertUsername.show();
	}
	
	/**
	 * Shows a notification with the specified title, message and intent class
	 * @param title
	 * @param message
	 * @param intentClass
	 */
	public static void notification(String title, String message, Class<?> intentClass)
	{
		notification("SteamDroid", title, message, intentClass, null, null);
	}
	
	/**
	 * Shows a notification with the specified parameter
	 * @param title
	 * @param ticker
	 * @param message
	 * @param intentClass
	 * @param key
	 * @param value
	 */
	public static void notification(String title, String ticker, String message, Class<?> intentClass, String key, String value)
	{
		if(manager != null && context != null && PreferenceManager.getDefaultSharedPreferences(SteamAlerts.getContext()).getBoolean("prefEnableNotifications", true))
		{
			Log.v("SteamDroid", "Sending notification...");
			
			int icon = R.drawable.notification_icon;
			CharSequence tickerText = ticker;
			long when = System.currentTimeMillis();
			Notification notification = new Notification(icon, tickerText, when);
			
			CharSequence contentTitle = "SteamDroid | " + title;
			CharSequence contentText = message;
			Intent intent = new Intent(context, intentClass);
			
			if(key != null && value != null)
			{
				intent.putExtra(key, value);
			}
			
			PendingIntent contentIntent = PendingIntent.getActivity(context, 0, intent, 0);
			
			notification.flags = Notification.FLAG_AUTO_CANCEL;
			notification.setLatestEventInfo(context, contentTitle, contentText, contentIntent);
			
			manager.notify(1, notification);
		}
	}
	
	/**
	 * Plays the message received sound
	 */
	public static void playSound()
	{
		if(PreferenceManager.getDefaultSharedPreferences(SteamAlerts.getContext()).getBoolean("prefEnableSounds", true))
		{
			mediaPlayer.start();
		}
	}
	
	/**
	 * Vibrates for the specified duration
	 * @param duration
	 */
	public static void vibrate(int duration)
	{
		if(PreferenceManager.getDefaultSharedPreferences(SteamAlerts.getContext()).getBoolean("prefEnableVibrate", true))
		{
			vibrator.vibrate(duration);
		}
	}
}

File: SteamDroid2/SteamDroid2/Util/SteamAlerts.cs
using System;
using Android.App;
using Android.Content;
using Android.Media;
using Android.OS;
using Android.Preferences;
using Android.Widget;
using Android.Util;

namespace SteamDroid.Util
{
    public class SteamAlerts
    {
        private static MediaPlayer mediaPlayer;

        private static AudioManager audioManager;
        private static NotificationManager notificationManager;
        private static Vibrator vibrator;
        
        private static Context context;
        
        private static ProgressDialog dialog;
        private static AlertDialog.Builder alert;
        
        private static bool isAlertsEnabled;
        private static bool isProgressEnabled;
        
        public static void Initialize(Activity activity)
        {
            audioManager = (AudioManager)activity.GetSystemService(Context.AudioService);
            
            mediaPlayer = MediaPlayer.Create(activity.ApplicationContext, Resource.Raw.Message);
            
            notificationManager = (NotificationManager)activity.GetSystemService(Context.NotificationService);
            vibrator = (Vibrator)activity.GetSystemService(Context.VibratorService);
            
            //context = activity.ApplicationContext;
            context = activity;

            isAlertsEnabled = true;
            isProgressEnabled = true;
        }
        
        public static Context GetContext()
        {
            return context;
        }
        
        public static void ShowAlertDialog(String title, String message, Context context)
        {
            HideProgressDialog();
            
            if(isAlertsEnabled)
            {
                AlertDialog.Builder dialog = new AlertDialog.Builder(context);
                dialog.SetTitle(title);
                dialog.SetMessage(message);
                dialog.SetPositiveButton("Ok", delegate (object sender, DialogClickEventArgs e) {
                    ((Dialog)sender).Cancel();
                });
                dialog.Show();
            }
        }
        
        public static void EnableAlerts()
        {
            isAlertsEnabled = true;
        }
        
        public static void DisableAlerts()
        {
            isAlertsEnabled = false;
        }

        public static void EnableProgress()
        {
            isProgressEnabled = true;
        }

        public static void DisableProgress()
        {
            isProgressEnabled = false;
            HideProgressDialog();
        }
        
        public static void ShowProgressDialog(String title, String message, Context context)
        {
            if(dialog == null)
            {
                dialog = new ProgressDialog(context);
            }
            
            dialog.SetTitle(title);
            dialog.SetMessage(message);
            dialog.Show();
        }
        
        public static void ShowInputDialog(String title, String message, TextView textView, EventHandler<DialogClickEventArgs> handler, Context context)
        {
            alert = new AlertDialog.Builder(context);
            alert.SetTitle(title);
            alert.SetMessage(message);
            
            alert.SetView(textView);
            
            alert.SetPositiveButton("Ok", handler);
            alert.Show();
        }
        
        public static void HideInputDialog()
        {
            if(dialog != null)
            {
                dialog.Cancel();
            }
        }

        public static void HideProgressDialog()
        {
            if (dialog != null)
            {
                dialog.Hide();
            }
        }
        
        public static void Notification(String title, String ticker, String message, Intent intent, String key, String value)
        {
            ISharedPreferences pref = PreferenceManager.GetDefaultSharedPreferences(context);
            if(pref.GetBoolean("prefEnableNotifications", true))
            {
                int icon = Resource.Drawable.NotificationIcon;
                Notification notification = new Notification(icon, ticker);
                
                if(key != null && value != null)
                {
                    intent.PutExtra(key, value);
                }
                
                PendingIntent contentIntent = PendingIntent.GetActivity(context, 0, intent, 0);

                bool permNotification = pref.GetBoolean("prefEnablePermNotification", false);

                if (!permNotification)
                {
                    notification.Flags = NotificationFlags.AutoCancel;
                }

                notification.SetLatestEventInfo(context, "SteamDroid | " + title, message, contentIntent);
                
                notificationManager.Notify(1, notification);
            }
        }


        public static void ShowToast(String message)
        {
            HideProgressDialog();

            Toast toast = Toast.MakeText(context, message, ToastLength.Short);
            toast.Show();
        }

        public static void Vibrate(int duration)
        {
            ISharedPreferences pref = PreferenceManager.GetDefaultSharedPreferences(context);
            if(pref.GetBoolean("prefEnableVibrate", true))
            {
                vibrator.Vibrate(duration);
            }
        }

        public static void PlaySound()
        {
            if (PreferenceManager.GetDefaultSharedPreferences(GetContext()).GetBoolean("prefEnableSound", true))
            {
                int streamVolume = audioManager.GetStreamVolume(Stream.Notification);
                int streamVolumeMax = audioManager.GetStreamMaxVolume(Stream.Notification);
                float volume = (float)streamVolume / (float)streamVolumeMax;
                mediaPlayer.SetVolume(volume, volume);

                mediaPlayer.Start();
            }
        }
    }
}

